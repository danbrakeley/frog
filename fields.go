package frog

import (
	"fmt"
	"strconv"
	"time"
)

type Field struct {
	IsJSONString bool // if true, the string in Value should be bookended by double quotes to be valid JSON
	IsJSONSafe   bool // if true, this string only contains alpha-numerics, spaces, and safe punctuation
	Name         string
	Value        string
}

// Fielder is an interface used to add structured logging to calls to Logger methods
type Fielder interface {
	Field() Field
}

// Bool adds a field whose value will be true or false
func Bool(name string, value bool) FieldBool {
	return FieldBool{Name: name, Value: value}
}

// Byte adds an 8-bit unsigned integer field
func Byte(name string, value byte) FieldUint64 {
	return FieldUint64{Name: name, Value: uint64(value)}
}

// Dur adds a time.Duration field
func Dur(name string, value time.Duration) FieldDuration {
	return Duration(name, value)
}

// Duration adds a time.Duration field
func Duration(name string, value time.Duration) FieldDuration {
	return FieldDuration{Name: name, Value: value}
}

// Err adds an error field named "error"
func Err(value error) FieldError {
	return FieldError{Name: "error", Value: value}
}

// Float32 adds a 32-bit floating point number field
func Float32(name string, value float32) FieldFloat64 {
	return FieldFloat64{Name: name, Value: float64(value)}
}

// Float64 adds a 64-bit floating point number field
func Float64(name string, value float64) FieldFloat64 {
	return FieldFloat64{Name: name, Value: value}
}

// Int adds a signed integer field
func Int(name string, value int) FieldInt64 {
	return FieldInt64{Name: name, Value: int64(value)}
}

// Int8 adds an 8-bit signed integer field
func Int8(name string, value int8) FieldInt64 {
	return FieldInt64{Name: name, Value: int64(value)}
}

// Int16 adds a 16-bit signed integer field
func Int16(name string, value int16) FieldInt64 {
	return FieldInt64{Name: name, Value: int64(value)}
}

// Int32 adds a 32-bit signed integer field
func Int32(name string, value int32) FieldInt64 {
	return FieldInt64{Name: name, Value: int64(value)}
}

// Int64 adds a 64-bit signed integer field
func Int64(name string, value int64) FieldInt64 {
	return FieldInt64{Name: name, Value: value}
}

// String adds an escaped and quoted string field
func String(name string, value string) FieldString {
	return FieldString{Name: name, Value: value}
}

// Time adds a time.Time field that will output a string formatted using RFC 3339 (ISO 8601)
func Time(name string, value time.Time) FieldTimeFormat {
	return FieldTimeFormat{Name: name, Value: value, Format: time.RFC3339}
}

// TimeNano adds a time.Time field that will output a string formatted using RFC 3339 with nanosecond precision
func TimeNano(name string, value time.Time) FieldTimeFormat {
	return FieldTimeFormat{Name: name, Value: value, Format: time.RFC3339Nano}
}

// TimeUnix adds a time.Time field that outputs as a unix epoch (unsigned integer)
func TimeUnix(name string, value time.Time) FieldTimeUnix {
	return FieldTimeUnix{Name: name, Value: value}
}

// TimeUnixNano adds a time.Time field that outputs as a unix epoch with nanosecond precision (unsigned integer)
func TimeUnixNano(name string, value time.Time) FieldTimeUnix {
	return FieldTimeUnix{Name: name, Value: value, Nano: true}
}

// Uint adds an unsigned integer field
func Uint(name string, value uint) FieldUint64 {
	return FieldUint64{Name: name, Value: uint64(value)}
}

// Uint8 adds an 8-bit unsigned integer field
func Uint8(name string, value uint8) FieldUint64 {
	return FieldUint64{Name: name, Value: uint64(value)}
}

// Uint16 adds a 16-bit unsigned integer field
func Uint16(name string, value uint16) FieldUint64 {
	return FieldUint64{Name: name, Value: uint64(value)}
}

// Uint32 adds a 32-bit unsigned integer field
func Uint32(name string, value uint32) FieldUint64 {
	return FieldUint64{Name: name, Value: uint64(value)}
}

// Uint64 adds a 64-bit unsigned integer field
func Uint64(name string, value uint64) FieldUint64 {
	return FieldUint64{Name: name, Value: value}
}

// Bool

type FieldBool struct {
	Name  string
	Value bool
}

func (f FieldBool) Field() Field {
	if f.Value {
		return Field{Name: f.Name, Value: "true"}
	}
	return Field{Name: f.Name, Value: "false"}
}

// Duration

type FieldDuration struct {
	Name  string
	Value time.Duration
}

func (f FieldDuration) Field() Field {
	return Field{IsJSONString: true, IsJSONSafe: true, Name: f.Name, Value: f.Value.String()}
}

// Error

type FieldError struct {
	Name  string
	Value error
}

func (f FieldError) Field() Field {
	if f.Value == nil {
		return Field{Name: f.Name, Value: "null"}
	}
	return Field{IsJSONString: true, Name: f.Name, Value: f.Value.Error()}
}

// Float32, Float64

type FieldFloat64 struct {
	Name  string
	Value float64
}

func (f FieldFloat64) Field() Field {
	return Field{Name: f.Name, Value: fmt.Sprintf("%g", f.Value)}
}

// Int, Int8, Int16, Int32, Int64

type FieldInt64 struct {
	Name  string
	Value int64
}

func (f FieldInt64) Field() Field {
	return Field{Name: f.Name, Value: strconv.FormatInt(f.Value, 10)}
}

// String

type FieldString struct {
	Name  string
	Value string
}

func (f FieldString) Field() Field {
	return Field{IsJSONString: true, IsJSONSafe: false, Name: f.Name, Value: f.Value}
}

// Time

type FieldTimeFormat struct {
	Name   string
	Value  time.Time
	Format string
}

func (f FieldTimeFormat) Field() Field {
	return Field{IsJSONString: true, IsJSONSafe: true, Name: f.Name, Value: f.Value.Format(f.Format)}
}

type FieldTimeUnix struct {
	Name  string
	Value time.Time
	Nano  bool
}

func (f FieldTimeUnix) Field() Field {
	if f.Nano {
		return Field{Name: f.Name, Value: strconv.FormatInt(f.Value.UnixNano(), 10)}
	}
	return Field{Name: f.Name, Value: strconv.FormatInt(f.Value.Unix(), 10)}
}

// Uint, Uint8, Uit16, Uint32, Uint64, Byte

type FieldUint64 struct {
	Name  string
	Value uint64
}

func (f FieldUint64) Field() Field {
	return Field{Name: f.Name, Value: strconv.FormatUint(f.Value, 10)}
}
