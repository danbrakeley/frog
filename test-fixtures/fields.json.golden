{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"bool","true":true}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"bool","false":false}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"byte","min":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"byte","max":255}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"time.Duration","how_long":"2m5s"}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"time.Duration","this_long":"4h48m1s"}
{"timestamp":"2019-09-10T21:44:00Z","level":"error","msg":"error","error":"this is the error"}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"error","error":null}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"float32","floatymc":3.333343267440796}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"float32","floatface":-2.0000000072549875e-15}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"float64","flargen":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"float64","blargen":-1.234456e+78}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"int","zero":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"int","negative":-1}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"int8","max":127}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"int8","min":-128}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"int16","max":32767}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"int16","min":-32768}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"int32","max":2147483647}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"int32","min":-2147483648}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"int64","max":9223372036854775807}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"int64","min":-9223372036854775808}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","empty":""}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","space":" "}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","quotes":"\""}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","newline":"\n"}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","newline":"a"}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"string","punctuation":"!@#$%^\u0026*()_+-=[]{}|;':,.\u003c\u003e?"}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"string","long":"this is a relatively long sentence with ʎzɐɹɔ cha\rac\ters i\n it \u0001 \"\u003c\u003c\u0026\u0026\u003e\u003e\""}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"time.Time","party":"1999-01-01T00:00:00Z"}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"time.Time","future":"2038-07-13T02:55:13Z"}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"time.Time (nano)","party":"1999-01-01T00:00:00Z"}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"time.Time (nano)","future":"2038-07-13T02:55:13.012398456Z"}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"time.Time (unix)","party":915148800}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"time.Time (unix)","future":2162602513}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"time.Time (unix,nano)","party":915148800000000000}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"time.Time (unix,nano)","future":2162602513012398456}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"uint","zero":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"uint","one":1}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"uint8","max":255}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"uint8","min":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"uint16","max":65535}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"uint16","min":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"uint32","max":4294967295}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"uint32","min":0}
{"timestamp":"2019-09-10T21:44:00Z","level":"info","msg":"uint64","max":18446744073709551615}
{"timestamp":"2019-09-10T21:44:00Z","level":"warning","msg":"uint64","min":0}
